### PR TITLE
S3 multypart upload

### PR DESCRIPTION
**Pull request by willyborankin**
The new solution for `S3OutputStream`
Reasons:
 Existing solution does not support S3 multy-part upload as  it is supposed to be:
 1. In the method `write` instead of creating the new Upload Part and push it to S3, buffer extended its size (using amortizing 
      algorithm). Since `S3OutputStream` internal buffer was not configurable there were situations in which connect service fails 
      with OOM exception
 2. `abort` method never used as result in case of OOM S3 contains 0 bytes files or corrupted one 
 
The new solution:
1. Introduces a new variable `aws.s3.part.size` with default size `5 MB` up to `2 GB` (no need to allocated `5 GB`)
2. Re-uses allocated buffer in `S3OutputStream` and flushes data into S3

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -138,6 +138,10 @@ dependencies {
     testImplementation "org.mockito:mockito-core:3.7.7"
 
     testRuntimeOnly "org.junit.jupiter:junit-jupiter-engine:$junitVersion"
+
+    testCompile "org.mockito:mockito-junit-jupiter:3.8.0"
+    testCompile "org.mockito:mockito-core:3.8.0"
```
Comment (AnatolyPopov) [inline build.gradle:143]: There is a dependency to mockito-core defined 5 lines above. 
Also `testCompile` seems to be deprecated in recent Gradle versions so I believe it should be `testImplementation`.
And could you please move the version to a variable like it's done for other libraries?

File: src/main/java/io/aiven/kafka/connect/s3/S3OutputStream.java [hunk 2]
```diff
@@ -17,97 +17,152 @@
...
+        }
+        if (byteBuffer.position() > 0 && Objects.nonNull(multipartUpload)) {
+            flushBuffer(byteBuffer.arrayOffset(), byteBuffer.position(), byteBuffer.position());
+        }
+        if (Objects.nonNull(multipartUpload)) {
+            multipartUpload.complete();
+            multipartUpload = null;
+        }
+        closed = true;
+        super.close();
```
Comment (AnatolyPopov) [inline src/main/java/io/aiven/kafka/connect/s3/S3OutputStream.java:114]: Seems that parent's `close` method is empty, so why this call is needed?

Comment (willyborankin) [reply src/main/java/io/aiven/kafka/connect/s3/S3OutputStream.java:114]: well I prefer to use it this way. Lets say a best practice

File: src/main/java/io/aiven/kafka/connect/s3/config/S3SinkConfig.java [hunk 3]
```diff
@@ -234,6 +236,41 @@ private static void addAwsConfigGroup(final ConfigDef configDef) {
...
+                        }
+                        if (number.longValue() > Integer.MAX_VALUE) {
+                            throw new ConfigException(
+                                    name,
+                                    value,
+                                    "Part size must be no more: " + Integer.MAX_VALUE + " bytes (~2GB)"
+                            );
+                        }
+                    }
+                },
```
Comment (AnatolyPopov) [inline src/main/java/io/aiven/kafka/connect/s3/config/S3SinkConfig.java:265]: Basically stealing the comment from another PR: Why not
```suggestion
        ConfigDef.Range.between(S3OutputStream.DEFAULT_PART_SIZE, Integer.MAX_VALUE),
```

Comment (willyborankin) [reply src/main/java/io/aiven/kafka/connect/s3/config/S3SinkConfig.java:265]: true true. i forget about `Range` will change

File: src/test/java/io/aiven/kafka/connect/s3/S3OutputStreamTest.java [hunk 4]
```diff
@@ -16,63 +16,314 @@
...
+    void noRequestsForEmptyBytes() throws Exception {
+
+        try (final var out = new S3OutputStream(BUCKET_NAME, FILE_KEY, 10, mockedAmazonS3)) {
+            out.write(new byte[] {});
+        }
+
+        verify(mockedAmazonS3, never()).initiateMultipartUpload(any(InitiateMultipartUploadRequest.class));
+        verify(mockedAmazonS3, never()).uploadPart(any(UploadPartRequest.class));
+        verify(mockedAmazonS3, never()).completeMultipartUpload(any(CompleteMultipartUploadRequest.class));
+        verify(mockedAmazonS3, never()).abortMultipartUpload(any(AbortMultipartUploadRequest.class));
```
Comment (AnatolyPopov) [inline src/test/java/io/aiven/kafka/connect/s3/S3OutputStreamTest.java:87]: Maybe `verifyNoMoreInteractions(mockedAmazonS3);` instead of this 4 lines

Comment (willyborankin) [reply src/test/java/io/aiven/kafka/connect/s3/S3OutputStreamTest.java:87]: It checks that life-circle is ok and all methods involved into it.

**Review by AnatolyPopov — DISMISSED** [review]

**Review by AnatolyPopov — APPROVED** [review]

**Pull request merged**